### PR TITLE
add support of http serving with inference model

### PR DESCRIPTION
**Pull request by Gavin-Sun555**
### 1. Add support of inference model HTTP serving.
- Add support of directly load inference model with "TensorflowModel" ，"OpenVINOModel"，“BigDL”， “Caffe”， “Pytorch”

### 2. Corresponding timers added
- Add timers "makeActivityTimer" to trigger the time of making activity from input data
- Add timers "handleResponseTimer" to trigger the time of handle responses

### 3. Performance
- Comparing with cluster serving, we test the performance for two sample models
The test is conducted in the same environment with tool wrk.

c: client/connection number
m: model/concurrent number
**On Node 200:**
m = 1
**Inference Model**

Model Type | DIEN | DIEN | DIEN | ResNet50 | ResNet50 | ResNet50
-- | -- | -- | -- | -- | -- | --
Command | wrk -t1 -c1 -d30s | wrk -t1 -c4 -d30s | wrk -t4 -c4 -d30s | wrk -t1 -c1 -d30s | wrk -t1 -c4 -d30s | wrk -t4 -c4 -d30s
wrk data | 18.94ms | 21.22ms | 21ms | 31.28ms | 90.57ms | 91.19ms

**Cluster Serving**

Model Type | DIEN | DIEN | DIEN | ResNet50 | ResNet50 | ResNet50
-- | -- | -- | -- | -- | -- | --
Command | wrk -t1 -c1 -d30s | wrk -t1 -c4 -d30s | wrk -t4 -c4 -d30s | wrk -t1 -c1 -d30s | wrk -t1 -c4 -d30s | wrk -t4 -c4 -d30s
wrk data | 64.5ms | 184.64ms | 185.8ms | 122.5ms | 290ms | 290ms

**On Clx 15:**
Our Inference Model Performance
n = batch of input size (# of instance)

scenario | n | c | m | predict time | backend time | e2e avg latency 
-- | -- | -- | -- | -- | -- | --
local clx015 | 1 | 1 | 1 | 14.54 | 17.12 | 17.89 
local clx015 | 1 | 2 | 1 | 22.51 | 23.32 | 24.25 
local clx015 | 1 | 1 | 2 | 13.58 | 16.89 | 17.68 
local clx015 | 1 | 1 | 4 | 23.10 | 25.94 | 26.10
local clx015 | 1 | 2 | 4 | 14.42 | 17.44 | 18.06
local clx015 | 1 | 4 | 4 | 15.80 | 19.01 | 20.02 
local clx015 | 2 | 1 | 1 | 12.04 | 29.61 | 30.40
local clx015 | 2 | 2 | 1 | 20.64 | 47.02 | 47.82 
local clx015 | 2 | 1 | 2 | 12.09 | 29.79 | 30.49 
local clx015 | 2 | 1 | 4 | 12.18 | 28.64 | 30.45
local clx015 | 2 | 2 | 4 | 13.12 | 30.62 | 31.13
local clx015 | 2 | 4 | 4 | 13.98 | 32.81 | 35.75
local clx015 | 4 | 1 | 1 | 14.57 | 69.42 | 71.50
local clx015 | 4 | 1 | 4 | 12.17 | 55.77 | 58.57
local clx015 | 4 | 4 | 4 | 14.19 | 64.35 | 66.21
local clx015 | 8 | 1 | 1 | 12.59 | 115.25 | 112.71
local clx015 | 16 | 1 | 1 | 14.18 | 250.34 | 251.03

Cluster Serving performance (Akka HTTP + Redis + Flink) provided by @EvelynQiang

scenario | c | m | predict time | backend time | e2e avg latency
-- | -- | -- | -- | -- | --
local clx015 | 1 | 1 | 18.94 | 20.42 | 25.51
local clx015 | 2 | 1 | 14.29 | 15.83 | 35.15
local clx015 | 1 | 2 | 13.09 | 14.47 | 17.66
local clx015 | 1 | 4 | 18.75 | 20.34 | 23.27
local clx015 | 2 | 4 | 14.75 | 16.16 | 18.57
local clx015 | 4 | 4 | 13.48 | 14.81 | 18.20

### CODE REVIEW

File: zoo/src/main/scala/com/intel/analytics/zoo/serving/http/domains.scala [hunk 1]
```diff
@@ -704,6 +803,47 @@ abstract class Servable(modelMetaData: ModelMetaData) {
...
+  extends Servable(inferenceModelMetaData) {
+  var model: InferenceModel = _
+
+  def load(): Unit = {
+    model = new InferenceModel(1)
+    inferenceModelMetaData.modelType match {
+      case "OpenVINO" =>
+        model.doLoadOpenVINO(inferenceModelMetaData.modelPath,
+          inferenceModelMetaData.weightPath)
+      case "frozenModel" =>
```
Comment (glorysdj) [inline zoo/src/main/scala/com/intel/analytics/zoo/serving/http/domains.scala:816]: need to add more support to other type of models
and may change frozenModel to tf.frozenModel

**Comment by glorysdj** [comment]
please also add some unit tests

**Comment by glorysdj** [comment]
please also add performance test data.

**Review by glorysdj — COMMENTED** [review]
we should re-organize the pom, especially scala and spark dependencies

File: zoo/pom.xml [hunk 2]
```diff
@@ -350,19 +350,19 @@
...
-        <!--            <groupId>org.apache.spark</groupId>-->
-        <!--            <artifactId>spark-core_${scala.major.version}</artifactId>-->
-        <!--            <version>${spark.version}</version>-->
-        <!--            <scope>${spark-scope}</scope>-->
-        <!--        </dependency>-->
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-core_${scala.major.version}</artifactId>
+            <version>${spark.version}</version>
+            <scope>${scala-library-scope}</scope>
```
Comment (glorysdj) [inline zoo/pom.xml:357]: please exclude it in related jar packages

File: zoo/pom.xml [hunk 3]
```diff
@@ -350,19 +350,19 @@
...
+        </dependency>
 
 
         <dependency>
             <groupId>org.apache.spark</groupId>
             <artifactId>spark-mllib_${scala.major.version}</artifactId>
             <version>${spark.version}</version>
-            <scope>${spark-scope}</scope>
+            <scope>${scala-library-scope}</scope>
```
Comment (glorysdj) [inline zoo/pom.xml:365]: please exclude it in related jar packages

**Comment by jason-dai** [comment]
What's the performance benefits? Please also compare to the Friesian performance result by @Litchilitchy

File: zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala [hunk 4]
```diff
@@ -85,52 +86,59 @@ object FrontEndApp extends Supportive with EncryptSupportive {
...
         }
+        logger.info("Servable Manager Load Success!")
       }
-      logger.info("Servable Manager Load Success!")
-
-      val rateLimiter: RateLimiter = arguments.tokenBucketEnabled match {
-        case true => RateLimiter.create(arguments.tokensPerSecond)
-        case false => null
-      }
+      var rateLimiter: RateLimiter = null
```
Comment (glorysdj) [inline zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala:95]: rateLimiter should keep 94-97 lines

File: zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala [hunk 5]
```diff
@@ -85,52 +86,59 @@ object FrontEndApp extends Supportive with EncryptSupportive {
...
+      var redisPutter: ActorRef = null
+      var redisGetter: ActorRef = null
+      var querierQueue: LinkedBlockingQueue[ActorRef] = null
```
Comment (glorysdj) [inline zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala:98]: redisPutter redisGetter and querierQueue are not needed

File: zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala [hunk 6]
```diff
@@ -85,52 +87,59 @@ object FrontEndApp extends Supportive with EncryptSupportive {
...
+      if (!arguments.multiServingMode) {
```
Comment (glorysdj) [inline zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala:100]: i think rateLimiter has nothing to do with multiServingMode

Comment (glorysdj) [reply zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala:100]: we can remove route ~ (post & path("predict")

File: zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala [hunk 7]
```diff
@@ -280,6 +289,9 @@ object FrontEndApp extends Supportive with EncryptSupportive {
           concat(
             (get & path(Segment)) {
               (modelName) => {
+                if (arguments.multiServingMode) {
+                  complete(404, "Serving Not In MultiServing Mode. Current Path not Supported")
```
Comment (glorysdj) [inline zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala:293]: and we can remove MultiServing Mode, we only provide a mode which is MultiServing mode

File: zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala [hunk 8]
```diff
@@ -321,27 +336,40 @@ object FrontEndApp extends Supportive with EncryptSupportive {
...
+                          val outputs = timing("model predict total")(modelInferenceTimer) {
                             servable.predict(instances)
                           }
                           Predictions(outputs)
                         }
                         timing("cluster serving response complete")() {
                           complete(200, result.toString)
                         }
+                      case _: InferenceModelServable =>
+                        val result = timing("inference model inference")(predictRequestTimer) {
```
Comment (glorysdj) [inline zoo/src/main/scala/com/intel/analytics/zoo/serving/http/FrontEndApp.scala:364]: these timers need to refactor
backend inference timer
model predict timer
etc

File: zoo/src/main/scala/com/intel/analytics/zoo/serving/http/test.scala [hunk 9]
```diff
@@ -0,0 +1,59 @@
+//package com.intel.analytics.zoo.serving.http
```
Comment (glorysdj) [inline zoo/src/main/scala/com/intel/analytics/zoo/serving/http/test.scala:1]: if you dont need this, delete it

**Comment by Gavin-Sun555** [comment]
Success Jenkins PR validation attached
http://10.239.176.111:18888/view/ZOO-PR/job/ZOO-PR-Validation/5871/

**Pull request merged**